### PR TITLE
Store security analysis status

### PR DESCRIPTION
**Pull request by ChamseddineBhd**

### CODE REVIEW

File: src/main/java/org/gridsuite/securityanalysis/server/repository/SecurityAnalysisResultRepository.java [hunk 1]
```diff
@@ -171,7 +175,8 @@ private static ContingencyEntity toEntity(UUID resultUuid, Contingency contingen
         Mono<Void> v1 = computationStatusRepository.deleteByResultUuid(resultUuid);
         Mono<Void> v2 = limitViolationRepository.deleteByResultUuid(resultUuid);
         Mono<Void> v3 = contingencyRepository.deleteByResultUuid(resultUuid);
-        return Flux.concat(v1, v2, v3)
+        Mono<Void> v4 = globalStatusRepository.deleteByResultUuid(resultUuid);
+        return Flux.concat(v1, v2, v3, v4)
```
Comment (jonenst) [inline src/main/java/org/gridsuite/securityanalysis/server/repository/SecurityAnalysisResultRepository.java:179]: In another PR, refactor this Flux.concat().then to Mono.when()

File: src/main/java/org/gridsuite/securityanalysis/server/repository/SecurityAnalysisResultRepository.java [hunk 2]
```diff
@@ -180,8 +185,24 @@ private static ContingencyEntity toEntity(UUID resultUuid, Contingency contingen
         Mono<Void> v1 = computationStatusRepository.deleteAll();
         Mono<Void> v2 = limitViolationRepository.deleteAll();
         Mono<Void> v3 = contingencyRepository.deleteAll();
-        return Flux.concat(v1, v2, v3)
+        Mono<Void> v4 = globalStatusRepository.deleteAll();
+        return Flux.concat(v1, v2, v3, v4)
```
Comment (jonenst) [inline src/main/java/org/gridsuite/securityanalysis/server/repository/SecurityAnalysisResultRepository.java:189]: In another PR, refactor this Flux.concat().then to Mono.when()

File: src/main/java/org/gridsuite/securityanalysis/server/service/SecurityAnalysisService.java [hunk 3]
```diff
@@ -39,8 +39,14 @@ public SecurityAnalysisService(SecurityAnalysisResultRepository resultRepository
     public Mono<UUID> runAndSaveResult(SecurityAnalysisRunContext runContext) {
         Objects.requireNonNull(runContext);
         UUID resultUuid = uuidGeneratorService.generate();
-        runPublisherService.publish(new SecurityAnalysisResultContext(resultUuid, runContext));
-        return Mono.just(resultUuid);
+
+        // update status to running status
+        return setStatus(resultUuid, SecurityAnalysisStatus.RUNNING.name()).then(
```
Comment (jonenst) [inline src/main/java/org/gridsuite/securityanalysis/server/service/SecurityAnalysisService.java:44]:
```java
return setStatus.then(
    mono.fromRunnable( () ->
       runPublisherService.publish(new SecurityAnalysisResultContext(resultUuid, runContext)
).thenReturn(resultUuid);
```
?

File: src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java [hunk 4]
```diff
@@ -118,4 +118,21 @@ private static SecurityAnalysisParameters getNonNullParameters(SecurityAnalysisP
         Mono<Void> result = service.deleteResults();
         return ResponseEntity.ok().body(result);
     }
+
+    @GetMapping(value = "/status/{resultUuid}", produces = APPLICATION_JSON_VALUE)
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java:122]: To be consistent with the other /results/{resultUuid}/status ?

File: src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java [hunk 5]
```diff
@@ -118,4 +118,21 @@ private static SecurityAnalysisParameters getNonNullParameters(SecurityAnalysisP
...
+    @Operation(summary = "Get the security analysis status from the database")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The security analysis status")})
+    public ResponseEntity<Mono<String>> getStatus(@Parameter(description = "Result UUID") @PathVariable("resultUuid") UUID resultUuid) {
+        Mono<String> result = service.getStatus(resultUuid);
+        return ResponseEntity.ok().body(result);
+    }
+
+    @PutMapping(value = "/status/{resultUuid}", produces = APPLICATION_JSON_VALUE)
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java:130]: To be consistent with the other /results/{resultUuid}/status ?

**Review by geofjamg — CHANGES_REQUESTED** [review]
See comment about API design

File: src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java (hunk 5 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java]: Corrected

File: src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java (hunk 4 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java]: Corrected

File: src/main/java/org/gridsuite/securityanalysis/server/service/SecurityAnalysisService.java (hunk 3 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/securityanalysis/server/service/SecurityAnalysisService.java:44]: Done

File: src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java [hunk 6]
```diff
@@ -118,4 +118,21 @@ private static SecurityAnalysisParameters getNonNullParameters(SecurityAnalysisP
...
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The security analysis status")})
+    public ResponseEntity<Mono<String>> getStatus(@Parameter(description = "Result UUID") @PathVariable("resultUuid") UUID resultUuid) {
+        Mono<String> result = service.getStatus(resultUuid);
+        return ResponseEntity.ok().body(result);
+    }
+
+    @PutMapping(value = "/results/{resultUuid}/status", produces = APPLICATION_JSON_VALUE)
+    @Operation(summary = "Set the security analysis status from the database")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The security analysis status has been set")})
+    public ResponseEntity<Mono<Void>> setStatus(@Parameter(description = "Result UUID") @PathVariable("resultUuid") UUID resultUuid,
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java:133]: We should remove this method as status is not supposed to be set from REST API.

**Review by geofjamg — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java (hunk 6 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java]: Changed to 'invalidateStatus'

File: src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java [hunk 7]
```diff
@@ -118,4 +118,20 @@ private static SecurityAnalysisParameters getNonNullParameters(SecurityAnalysisP
...
+
+    @GetMapping(value = "/results/{resultUuid}/status", produces = APPLICATION_JSON_VALUE)
+    @Operation(summary = "Get the security analysis status from the database")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The security analysis status")})
+    public ResponseEntity<Mono<String>> getStatus(@Parameter(description = "Result UUID") @PathVariable("resultUuid") UUID resultUuid) {
+        Mono<String> result = service.getStatus(resultUuid);
+        return ResponseEntity.ok().body(result);
+    }
+
+    @PutMapping(value = "/results/{resultUuid}/invalidateStatus", produces = APPLICATION_JSON_VALUE)
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java:130]: /results/{resultUuid}/invalidate-status ? snake case seems to be usual for REST API

Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/securityanalysis/server/SecurityAnalysisController.java]: Changed

**Review by geofjamg — APPROVED** [review]

**Pull request merged**